src/Blocks/BlockPiston.cpp

```diff
@@ -86,28 +73,129 @@ bool cBlockPistonHandler::GetPlacementBlockTypeMeta(
+	cWorld * a_World, const Vector3i & a_PushDir
+)
+{
+	// Sort blocks to move the blocks first, which are farest away from the piston
+	// This prevents the overwriting of existing blocks
+	std::vector<Vector3i> sortedBlocks(a_BlocksToPush.begin(), a_BlocksToPush.end());
+	std::sort(sortedBlocks.begin(), sortedBlocks.end(), [a_PushDir](const Vector3i & a, const Vector3i & b)
+	{
+		return a.Dot(a_PushDir) > b.Dot(a_PushDir);
+	});
+
+	// Move every block
+	BLOCKTYPE moveBlock;
+	NIBBLETYPE moveMeta;
+	for (Vector3i & moveBlockPos : sortedBlocks)
+	{
+		a_World->GetBlockTypeMeta(moveBlockPos.x, moveBlockPos.y, moveBlockPos.z, moveBlock, moveMeta);
+		a_World->SetBlock(moveBlockPos.x, moveBlockPos.y, moveBlockPos.z, E_BLOCK_AIR, 0);
+
+		moveBlockPos += a_PushDir;
+		if (cBlockInfo::IsPistonBreakable(moveBlock))
+		{
+			// Block is breakable, drop it
+			cBlockHandler * Handler = BlockHandler(moveBlock);
+			if (Handler->DoesDropOnUnsuitable())
+			{
+				cChunkInterface ChunkInterface(a_World->GetChunkMap());
+				cBlockInServerPluginInterface PluginInterface(*a_World);
+				Handler->DropBlock(ChunkInterface, *a_World, PluginInterface, nullptr,
+					moveBlockPos.x, moveBlockPos.y, moveBlockPos.z
+				);
+			}
+		} else
 		{
-			// This block cannot be pushed at all, the piston can't extend
-			return -1;
+			// Not breakable, just move it
+			a_World->SetBlock(moveBlockPos.x, moveBlockPos.y, moveBlockPos.z, moveBlock, moveMeta);
 		}
 	}
-	// There is no space for the blocks to move, piston can't extend
-	return -1;
+}
+
+
+
+
+
+bool cBlockPistonHandler::CanPushBlock(
+	int a_BlockX, int a_BlockY, int a_BlockZ, cWorld * a_World, bool a_RequirePushable,
+	std::unordered_set<Vector3i, VectorHasher<int>> & a_BlocksPushed, const Vector3i & a_PushDir
+)
+{
+	const static std::array<Vector3i, 6> pushingDirs = {{ Vector3i(-1, 0, 0), Vector3i(1, 0, 0), Vector3i(0, -1, 0), Vector3i(0, 1, 0),
+		Vector3i(0, 0, -1), Vector3i(0, 0, 1) }};
+
+	BLOCKTYPE currBlock;
+	NIBBLETYPE currMeta;
+	a_World->GetBlockTypeMeta(a_BlockX, a_BlockY, a_BlockZ, currBlock, currMeta);
+
+	if (currBlock == E_BLOCK_AIR)
+	{
+		// Air can be pushed
+		return true;
+	}
+
+	if (!a_RequirePushable && cBlockInfo::IsPistonBreakable(currBlock))
+	{
+		// Block should not be broken, when it's not in the pushing direction
+		return true;
+	}
+
+	if (!CanPush(currBlock, currMeta))
+	{
+		// When it's not required to push this block, don't fail
+		return !a_RequirePushable;
+	}
+
+	if (a_BlocksPushed.size() >= PISTON_MAX_PUSH_DISTANCE)
+	{
+		// Do not allow to push too much blocks
+		return false;
+	}
+
+	if (!a_BlocksPushed.emplace(a_BlockX, a_BlockY, a_BlockZ).second || cBlockInfo::IsPistonBreakable(currBlock))
+	{
+		return true;  // Element exist already
+	}
+
+	if (currBlock == E_BLOCK_SLIME_BLOCK)
+	{
+		// Try to push the other directions
+		for (const Vector3i & testDir : pushingDirs)
```

Comment:
`auto` is preferred because of easy refactoring. All new loops should be written with `auto` and range-based, if possible.